Improve error displayed to customers when an item's stock status changes during checkout.

opr opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
### Description
This PR adds some new exceptions that can be thrown when validating the cart contents. These exceptions all relate to different types of "out of stock"

##### Thrown in `validate_cart_item`
- `PartialOutOfStockException` - Thrown when there _are_ items in stock, but not enough to satisfy the customer's order. The quantity in cart will be reduced to the in-stock amount.
- `TooManyInCartException` - Thrown when products can only be purchased individually but there is more than one in the cart. The quantity in cart will be reduced to 1.
- `OutOfStockException` - Thrown when the product is completely out of stock. The item is removed from cart.
- `NotPurchasableException` - Thrown when the product is not able to be purchased (doesn't exist, doesn't have a price etc.)
##### Thrown in `validate_cart_items`
- `StockValidationException` - Thrown as a "general" exception from `validate_cart_items` when one/some/all of the above exceptions are thrown in `validate_cart_item`. This Exception contains a `WP_Error` object that in turn contains all of the errors thrown by each execution of`validate_cart_item`. This `WP_Error` is then converted into a response and sent to the client so that the individual error notices might be displayed.

<!-- Reference any related issues or PRs here -->
Fixes #3412
This supersedes #3469 

<!-- Don't forget to update the title with something descriptive. -->
<!-- If your pull request implements a feature flag, make sure you update [this doc](../docs/blocks/features-and-blocks-behind-a-flag.md) -->

### Screenshots

### Before
<img width="600" alt="Screenshot 2021-01-06 at 16 06 36" src="https://user-images.githubusercontent.com/5656702/103790580-4c655780-5039-11eb-96c1-ac78055f18f2.png">


### After
<img width="600" alt="Screenshot 2021-01-06 at 14 45 04" src="https://user-images.githubusercontent.com/5656702/103781101-c8f23900-502d-11eb-94bb-998d73373663.png">

### How to test the changes in this Pull Request:

1. **In an incognito window** Add several products to your cart. Add these in differing quantities. Several of these added products need to have a quantity of 3+ in the cart.
2. Open the checkout page and ensure all the products are showing. **Do different combinations of the following steps in a different session to the one where you've added the items, and also without refreshing the checkout page. This is necessary because we can't be logged in as admin when testing step 3 below.**
3. Delete one or more of the products from the back end.
4. Set one or more of the products to "Manage stock?" true, and edit the in-stock amount of one or more of the products ensuring it is below the quantity that is in the cart, but above 0.
5. Set one or more of the products to "Sold individually"
6. Set one or more of the products to be entirely out of stock.
7. Try to check out and ensure you see the correct error messages for the products whose stock you've changed/reduced.
8. Try a few different scenarios with stock levels, number of products affected, number of products in cart etc. 

<!-- If you can, add the appropriate labels -->

### Changelog

> Improve error displayed to customers when an item's stock status changes during checkout.

opr committed:
Add new exceptions for out of stock scenarios

These are needed to differentiate between the different stock validation errors, and so we can create the correct error message.

opr committed:
Catch new out of stock exceptions when checking the cart for errors

This is so we can get the cart sent back to the client, if we don't catch these, then the route will just return a 500 error and crash.

opr committed:
Add ArrayUtils class

This will contain methods used to operate on arrays that don't fit anywhere else.

opr committed:
Handle the case in Checkout where the error is already a WP_Error

This will happen when the cart fails validation.

opr committed:
Handle StockAvailabilityException in AbstractRoute

This will happen when an item or number of items in the cart are out of stock/insufficient stock.

opr committed:
Throw exceptions for each type of invalid stock in validate_cart_items

This will allow us to create an error message for each type of violation to display to the user.

opr committed:
Display additional error notices returned by the API

opr committed:
Fix wording when throwing exceptions relating to stock

opr committed:
Handle TooManyInCartException in CartController

mikejolley commented on src/StoreApi/Routes/AbstractRoute.php:
```diff
@@ -72,6 +73,8 @@ public function get_response( \WP_REST_Request $request ) {
 		} catch ( RouteException $error ) {
 			$response = $this->get_route_error_response( $error->getErrorCode(), $error->getMessage(), $error->getCode(), $error->getAdditionalData() );
+		} catch ( StockAvailabilityException $error ) {
+			$response = $this->get_route_error_response( $error->getErrorCode(), $error->getError(), $error->getCode(), $error->getAdditionalData() );
```
We may need to replace or refactor `get_route_error_response` since it's inherited from the Abstract and we're changing the params here (string to WP_Error Object).

Perhaps a new method is needed? `get_route_error_response_from_object`?

opr replied:
Good call, I wasn't sure if that was overkill, so I didn't do it, but since you suggested it too I have implemented it.

mikejolley commented on src/StoreApi/Routes/Checkout.php:
```diff
@@ -277,6 +282,19 @@ protected function get_route_error_response( $error_code, $error_message, $http_
 				$controller = new CartController();
 				$cart       = $controller->get_cart_instance();
 
+				if ( is_wp_error( $error_message ) ) {
```
If we go with the `get_route_error_response_from_object ` suggestion in another comment, we should abstract this chunk of code out and use it in both; appending the cart data should be shared between both.

opr replied:
Thanks for the suggestion, I've now added `add_data_to_error_object` to `Checkout` I don't think it needs to be defined in `AbstractRoute` because it's not used anywhere else except the Checkout. If other routes start returning `WP_Error` objects then we can move it but I think that's out of scope for this PR.

mikejolley commented on src/StoreApi/Routes/Checkout.php:
```diff
@@ -253,15 +255,18 @@ protected function get_route_post_response( WP_REST_Request $request ) {
-	 * @param int    $http_status_code HTTP status. Defaults to 500.
-	 * @param array  $additional_data  Extra data (key value pairs) to expose in the error response.
+	 * @param string          $error_code String based error code.
+	 * @param string|WP_Error $error_message User facing error message.
```
I think we should only handle strings in this method to avoid complexity.

mikejolley commented on src/StoreApi/Utilities/CartController.php:
```diff
@@ -3,7 +3,9 @@
 use Automattic\WooCommerce\Blocks\StoreApi\Utilities\NoticeHandler;
+use Automattic\WooCommerce\Blocks\Utils\ArrayUtils;
 use Automattic\WooCommerce\Checkout\Helpers\ReserveStock;
+use WP_Error;
```
Looks like all the backslashes in the rest of the codebase are unnecessary 🤦🏻 👍🏻

mikejolley commented on src/StoreApi/Utilities/CartController.php:
```diff
@@ -344,6 +463,12 @@ public function get_cart_item_errors() {
 			} catch ( RouteException $error ) {
 				$errors[] = new \WP_Error( $error->getErrorCode(), $error->getMessage() );
+			} catch ( NotPurchasableException $error ) {
+				$errors[] = new \WP_Error( $error->getErrorCode(), $error->getMessage() );
```
`\WP_Error` can just be `WP_Error` because its imported with the use operator.

mikejolley commented on src/StoreApi/Utilities/NotPurchasableException.php:
```diff
@@ -0,0 +1,73 @@
+		$this->error_code      = $error_code;
+		$this->product_name    = $product_name;
+		$this->additional_data = array_filter( (array) $additional_data );
+		parent::__construct();
```
Should we pass params to the parent constructor here?

mikejolley commented on src/StoreApi/Utilities/OutOfStockException.php:
```diff
@@ -0,0 +1,73 @@
+		$this->error_code      = $error_code;
+		$this->product_name    = $product_name;
+		$this->additional_data = array_filter( (array) $additional_data );
+		parent::__construct();
```
Should we pass params to the parent constructor here?

mikejolley commented on src/StoreApi/Utilities/PartialOutOfStockException.php:
```diff
@@ -0,0 +1,73 @@
+		$this->error_code      = $error_code;
+		$this->product_name    = $product_name;
+		$this->additional_data = array_filter( (array) $additional_data );
+		parent::__construct();
```
Should we pass params to the parent constructor here?

mikejolley commented on src/StoreApi/Utilities/NotPurchasableException.php:
```diff
@@ -0,0 +1,73 @@
+ * @internal This API is used internally by Blocks, this exception is thrown when an item in the cart is not able to be
+ * purchased.
+ */
+class NotPurchasableException extends \Exception {
```
All of these new Exception classes appear to be almost identical—perhaps we should have that code in one class and extend it? It may make catching the exceptions more straightforward too (if you can catch the parent class, would have to test to ensure this works).

opr replied:
The reason the exceptions are separate like this is because we need to catch them separately in `validate_cart_items` so we can generate the correct error messages. I agree that we should create a new exception to reduce code duplication in these exceptions, but I don't think we can catch the parent exception for the reason stated above.

I propose we rename what is currently `StockAvailabilityException` which is the **ultimate** exception, it contains the `WP_Error` object with all of the error messages inside, to something like `InvalidQuantitiesInCartException`. Then we can use the name `StockAvailabilityException` as the parent of the `OutOfStockException`, `PartialOutOfStockException`, `NotPurchasableException`, and `TooManyInCartException` exceptions.

mikejolley commented on src/StoreApi/Utilities/CartController.php:
```diff
@@ -288,20 +416,11 @@ public function validate_cart_item( $cart_item ) {
-					),
-					400
+
+				WC()->cart->set_quantity( $cart_item['key'], $qty_remaining, false );
```
Is this desirable every time the validate function runs on all requests, or should it be it's own method/optional? 

I'm not saying this is wrong, I'm just wondering if this is something we definitely want to do in all use cases or if we want to enable this somehow. I guess that might impact the error notices though..feel free to ignore.

opr replied:
I actually don't believe this is needed at all. Removing it now.

mikejolley commented on src/StoreApi/Utilities/CartController.php:
```diff
@@ -213,17 +215,144 @@ public function validate_add_to_cart( \WC_Product $product, $request ) {
 	 *
-	 * @throws RouteException Exception if invalid data is detected.
+	 * @throws StockAvailabilityException Exception if invalid data is detected due to insufficient stock levels.
+	 * @throws RouteException Exception if invalid data is detected but not due to insufficient stock levels.
```
Since this is thrown outside of this method, I think you can remove this line and the phpcs exception. It's documented within convert_notices_to_exceptions

mikejolley commented on src/StoreApi/Utilities/CartController.php:
```diff
@@ -213,17 +215,144 @@ public function validate_add_to_cart( \WC_Product $product, $request ) {
+					sprintf(
+						/* translators: %s: product names  */
+						_n(
+							'"%s" is out of stock and cannot be purchased.',
```
I like what you've done with these messages but one change needed would be to wrap each product name with `""` rather than the full list. e.g. "Product 1 and Product 2" would become "Product 1" and "Product 2".

opr replied:
Cheers, I've added this functionality to `natural_language_join`: https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/8094d9f98e7459e79d89c83eb116db9d82508fba/src/Utils/ArrayUtils.php#L17-L24

mikejolley commented on src/StoreApi/Utilities/CartController.php:
```diff
@@ -213,17 +215,144 @@ public function validate_add_to_cart( \WC_Product $product, $request ) {
+			}
+		}
+
+		if (
```
I think this section can be streamlined. If you create the WP_Error object regardless you can evaluate each error type separately, each using `$error->add` as needed, and then finally you would only need to do a check like this:

```
if ( $error->has_errors() ) {
			WC()->cart->calculate_totals();
			throw new StockAvailabilityException(
				'woocommerce_stock_availability_error',
				$error
			);
}
```

mikejolley commented on src/StoreApi/Utilities/CartController.php:
```diff
@@ -213,17 +215,144 @@ public function validate_add_to_cart( \WC_Product $product, $request ) {
+
+			$error = new WP_Error();
+
+			if ( count( $out_of_stock_products ) > 0 ) {
+				$out_of_stock_product_names = array_map(
+					function( $product ) {
+						return $product->getProductName();
+					},
+					$out_of_stock_products
+				);
+
+				$error->add(
+					409,
+					sprintf(
+						/* translators: %s: product names  */
+						_n(
+							'"%s" is out of stock and cannot be purchased.',
+							'"%s" are out of stock and cannot be purchased.',
+							count( $out_of_stock_products ),
+							'woo-gutenberg-products-block'
+						),
+						ArrayUtils::natural_language_join( $out_of_stock_product_names )
+					)
+				);
```
This appears to be repeated for each error type, the only difference is the error message text. This seems like it could be easily extracted into a separate method. e.g. `product_stock_error_message( $singular, $plural, $product_names )`. Just an idea, what you have works.

opr replied:
Yup I like the idea, thanks Mike, I've implemented `add_product_names_to_message` which takes `$singular_error`, `$plural_error`, and `$items`, returning a string.

mikejolley commented on src/StoreApi/Utilities/CartController.php:
```diff
@@ -213,17 +215,144 @@ public function validate_add_to_cart( \WC_Product $product, $request ) {
+			}
+
+			if ( count( $not_purchasable_products ) > 0 ) {
+				$not_purchasable_product_names = array_map(
```
You can use https://github.com/woocommerce/woocommerce/blob/686872a2372a71ed35b90a8ff790b046d0b40118/includes/wc-core-functions.php#L2055 if you like: `wc_list_pluck( $not_purchasable_products, 'getProductName' )`

mikejolley reviewed (commented):
Nice job @opr — this appears to be working well on first try! I have added some inline comments, most are minor. Once addressed I can do another round of testing.

opr committed:
Abstract the merging of cart, status, and additional data into new fn

This allows us to simplify the way errors are returned from the API. The reason we have to add all of the data at once is because of how WP_Error works with the additional data, if there is already existing data in a WP_Error object, it gets moved into additional_data. By adding all of the data in one place, we stop this from happening. Also since we're only adding status and/or cart explicitly, it makes sense to just do it in one place.

opr committed:
Add get_route_error_response_from_object method

This is so we can differentiate between a string and WP_Error object.

opr committed:
Remove unnecessary slashes from WP_Error instantiation

opr committed:
Add option to enclose each item in quotes in natural_language_join

opr committed:
Abstract adding error messages to error object into single function

A lot of code was repeated, so doing this cuts down on that and ensures any changes only need to be made in one place.

opr committed:
Create new parent exception for each type of out of stock exception

This is so we don't have to repeat code inside each different exception and we can simply inherit StockAvailabilityException.

opr committed:
Catch the generic StockAvailabilityException in get_cart_item_errors

opr committed:
No longer recalculate totals in validate function

It is not needed, the totals are recalculated elsewhere. This call was superfluous.

opr committed:
Reduce nesting, and only throw exception if error object has errors

mikejolley commented on src/StoreApi/Routes/AbstractRoute.php:
```diff
@@ -204,6 +208,19 @@ protected function get_route_error_response( $error_code, $error_message, $http_
 	}
 
+	/**
+	 * Get route response when something went wrong and the supplied error is a WP_Error. This happens when
```
> This happens when

Stray comment here.

mikejolley commented on src/StoreApi/Utilities/CartController.php:
```diff
@@ -213,17 +215,130 @@ public function validate_add_to_cart( \WC_Product $product, $request ) {
+				$this->add_product_names_to_message( $singular_error, $plural_error, $partial_out_of_stock_products )
+			);
+
+			if ( $error->has_errors() ) {
```
I think this is nested too deep? It's within the `if ( count( $partial_out_of_stock_products ) > 0 ) {` conditon.

opr replied:
Good spot, I didn't see this when testing because I had partial out of stock products.

mikejolley reviewed (changes requested):
Found a few small issues; I'll do a final test once those are fixed :) Thanks!

opr committed:
Improve comment on get_route_error_response_from_object method

opr committed:
Fix nesting when throwing the InvalidStockLevelsInCartException

mikejolley reviewed (approved):
Thanks for taking care of feedback. LGTM now 👍🏻  Nice job!
